lib/experimental/class-wp-html-processor.php

```diff
@@ -0,0 +1,720 @@
+ * eagerly allocate and convert string values we don't ever read.
+ *
+ * Classes are passed by reference by default, but are we missing
+ * opportunities to pass by reference when we are copying by value?
+ */
+
+
+/**
+ * Processes an input HTML document by applying a specified set
+ * of patches to that input. Tokenizes HTML but does not fully
+ * parse the input document.
+ */
+class WP_HTML_Processor {
+	/**
+	 * Describes the constraints imposed on the HTML tag we're trying to match.
+	 *
+	 * @see WP_Tag_Find_Descriptor::parse()
+	 *
+	 * @var WP_Tag_Find_Descriptor
+	 */
+	public $descriptor;
+
+
+	/**
+	 * Scans the HTML document, keeping track of its own state.
+	 *
+	 * @var WP_HTML_Scanner
+	 */
+	public $scanner;
+
+
+	/**
+	 * Set of modifications to perform on input HTML document, built
+	 * via calls to the fluid API exposed by WP_HTML_Processor.
+	 *
+	 * @var WP_Tag_Modifications
+	 */
+	public $modifications;
+
+
+	public function __construct( $input ) {
+		$this->scanner = new WP_HTML_Scanner( $input );
+		$this->modifications = new WP_Tag_Modifications();
+	}
+
+
+	public function find( $query ) {
+		$this->commit_class_changes();
+		$this->descriptor = WP_Tag_Find_Descriptor::parse( $query );
+		if ( false === $this->scanner->scan( $this->descriptor ) ) {
+			return false;
+		}
+
+		return $this;
+	}
+
+
+	public function add_attribute( $name, $value ) {
+		$tag_name = $this->scanner->tag_name;
+		$insert_at = $tag_name->start + $tag_name->length;
+
+		$this->modifications->replacements[] = array(
+			$insert_at,
+			$insert_at,
+			' ' . self::serialize_attribute( $name, $value )
+		);
+
+		return $this;
+	}
+
+
+	public function add_class( $class_name ) {
+		$this->modifications->class_changes[ $class_name ] = true;
+	}
+
+
+	public function set_attribute( $name, $value ) {
+		if ( ! array_key_exists( WP_Tag_Find_Descriptor::comparable( $name ), $this->scanner->attributes ) ) {
+			return $this->add_attribute( $name, $value );
+		}
+
+		/** @var WP_HTML_Attribute_Token $existing */
+		$existing = $this->scanner->attributes[ $name ];
+
+		$this->modifications->replacements[] = array(
+			$existing->attribute_start,
+			$existing->attribute_end,
+			self::serialize_attribute( $name, $value )
+		);
+	}
+
+
+	public function remove_class( $class_name ) {
+		$this->modifications->class_changes[ $class_name ] = false;
+	}
+
+
+	public function remove_attribute( $name ) {
+		if ( ! array_key_exists( $name, $this->scanner->attributes ) ) {
+			return $this;
+		}
+
+		/** @var WP_HTML_Attribute_Token $existing */
+		$existing = $this->scanner->attributes[ $name ];
+
+		$this->modifications->replacements[] = array(
+			$existing->attribute_start,
+			$existing->attribute_end,
+			''
+		);
+	}
+
+
+	public static function serialize_attribute( $name, $value ) {
+		$value = str_replace( '"', "&quot;", $value );
+
+		return "{$name}=\"{$value}\"";
+	}
+
+
+	public function commit_class_changes() {
```

Comment:
```suggestion
	private function commit_class_changes() {
```

This doesn't have to be exposed as a public method.